patches/net/minecraft/world/level/block/entity/HopperBlockEntity.java.patch

```diff
@@ -1,22 +1,88 @@
-+        if (net.neoforged.neoforge.items.VanillaInventoryCodeHooks.insertHook(p_326256_)) return true;
-         Container container = getAttachedContainer(p_155563_, p_155564_, p_326256_);
-         if (container == null) {
+-        Container container = getAttachedContainer(p_155563_, p_155564_, p_326256_);
+-        if (container == null) {
++        var containerOrHandler = getContainerOrHandlerAt(p_155563_, p_155564_.relative(p_326256_.facing), p_326256_.facing.getOpposite());
++        if (containerOrHandler.isEmpty()) {
              return false;
-@@ -214,6 +_,8 @@
+-        } else {
++        } else if (containerOrHandler.container() != null) {
++            Container container = containerOrHandler.container();
+             Direction direction = p_326256_.facing.getOpposite();
+             if (isFullContainer(container, direction)) {
+                 return false;
+@@ -164,6 +_,8 @@
+ 
+                 return false;
+             }
++        } else {
++            return net.neoforged.neoforge.items.VanillaInventoryCodeHooks.insertHook(p_326256_, containerOrHandler.itemHandler());
+         }
+     }
+ 
+@@ -214,8 +_,9 @@
      public static boolean suckInItems(Level p_155553_, Hopper p_155554_) {
          BlockPos blockpos = BlockPos.containing(p_155554_.getLevelX(), p_155554_.getLevelY() + 1.0, p_155554_.getLevelZ());
          BlockState blockstate = p_155553_.getBlockState(blockpos);
-+        Boolean ret = net.neoforged.neoforge.items.VanillaInventoryCodeHooks.extractHook(p_155553_, p_155554_);
-+        if (ret != null) return ret;
-         Container container = getSourceContainer(p_155553_, p_155554_, blockpos, blockstate);
-         if (container != null) {
+-        Container container = getSourceContainer(p_155553_, p_155554_, blockpos, blockstate);
+-        if (container != null) {
++        var containerOrHandler = getSourceContainerOrHandler(p_155553_, p_155554_, blockpos, blockstate);
++        if (containerOrHandler.container() != null) {
++            Container container = containerOrHandler.container();
              Direction direction = Direction.DOWN;
+ 
+             for (int i : getSlots(container, direction)) {
+@@ -225,6 +_,8 @@
+             }
+ 
+             return false;
++        } else if (containerOrHandler.itemHandler() != null) {
++            return net.neoforged.neoforge.items.VanillaInventoryCodeHooks.extractHook(p_155554_, containerOrHandler.itemHandler());
+         } else {
+             boolean flag = p_155554_.isGridAligned()
+                 && blockstate.isCollisionShapeFullBlock(p_155553_, blockpos)
+@@ -368,6 +_,8 @@
+         return p_155590_.getEntitiesOfClass(ItemEntity.class, aabb, EntitySelector.ENTITY_STILL_ALIVE);
+     }
+ 
++    /** @deprecated Use IItemHandler capability instead. To preserve Container-specific interactions, use {@link #getContainerOrHandlerAt} and handle both cases. */
++    @Deprecated
```

Comment:
Why was only this method deprecated? Why not the other `getContainerAt` overload, or `getSourceContainer`, or other nearby methods that work with `Container`?

Reply:
I only deprecated the public one, but I can deprecate all of them if you want.

Reply:
Ah right, the other methods are `private`. Do we think it is likely that a mod might access-transform any of those methods to being `public` for their use? If not, we can just deprecate the `public` one and leave it at that.

Reply:
I think it's unlikely that a modder would access-transform another of these methods without noticing that Minecraft doesn't use them anymore, and that the `public` `getContainerAt` overload is deprecated.